File: src/main/java/com/everyonewaiter/global/config/AsynchronousConfiguration.java
package com.everyonewaiter.global.config;

import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.task.TaskExecutor;
import org.springframework.core.task.support.TaskExecutorAdapter;
import org.springframework.scheduling.annotation.EnableAsync;

@Configuration
@EnableAsync
class AsynchronousConfiguration {

  @Bean(name = "eventTaskExecutor")
  public TaskExecutor eventTaskExecutor() {
    ThreadFactory factory = Thread.ofVirtual().name("app-event-", 1).factory();
    return new TaskExecutorAdapter(Executors.newThreadPerTaskExecutor(factory));
  }

}

File: src/main/kotlin/com/everyonewaiter/global/config/AsynchronousConfiguration.kt
package com.everyonewaiter.global.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.util.concurrent.ThreadPoolExecutor

@EnableAsync
@Configuration
class AsynchronousConfiguration {
    @Bean
    fun asyncTaskExecutor() =
        ThreadPoolTaskExecutor().apply {
            corePoolSize = 20
            maxPoolSize = 50
            queueCapacity = 100
            setThreadNamePrefix("app-event-")
            setRejectedExecutionHandler(ThreadPoolExecutor.CallerRunsPolicy())
            initialize()
        }
}
